sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,357 @@
+          synchronized (flushLock) {
+            receiver.accept(timers);
+            if (totalByteCounter.longValue() > sizeLimit) {
+              flush();
+            }
+          }
+        };
+      }
+    }
+    outputTimersReceivers.put(timerKey, receiver);
+    return timers -> {
+      receiver.accept(timers);
+      if (totalByteCounter.longValue() > sizeLimit) {
+        flush();
+      }
+    };
+  }
+
+  public void flush() throws IOException {
+    Elements.Builder elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    if (totalByteCounter.longValue() == 0
+        && outputTimersReceivers.isEmpty()
+        && outputDataReceivers.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    LOG.debug(
+        "Closing streams for outbound endpoints {} and {}.",
+        outputDataReceivers.keySet(),
+        outputTimersReceivers.keySet());
+    LOG.debug(
+        "Sent outbound data size : {}, outbound timers size : {}.",
+        outputDataReceivers.entrySet().stream()
+            .map(kv -> new AbstractMap.SimpleEntry<>(kv.getKey(), kv.getValue().localByteCount))
+            .collect(Collectors.toSet()),
+        outputTimersReceivers.entrySet().stream()
+            .map(kv -> new AbstractMap.SimpleEntry<>(kv.getKey(), kv.getValue().localByteCount))
+            .collect(Collectors.toSet()));
+    LOG.debug("Sent outbound element count : {}.", perEndpointElementCount);
+    for (String pTransformID : outputDataReceivers.keySet()) {
+      bufferedElements
+          .addDataBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(pTransformID)
+          .setIsLast(true);
+    }
+    for (KV<String, String> pTransformAndTimerId : outputTimersReceivers.keySet()) {
+      bufferedElements
+          .addTimersBuilder()
+          .setInstructionId(processBundleRequestIdSupplier.get())
+          .setTransformId(pTransformAndTimerId.getKey())
+          .setTimerFamilyId(pTransformAndTimerId.getValue())
+          .setIsLast(true);
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputDataReceivers.values().forEach(Receiver::reset);
+    outputTimersReceivers.values().forEach(Receiver::reset);
+    perEndpointElementCount.clear();
+  }
+
+  private Elements.Builder convertBufferForTransmission() {
+    Elements.Builder bufferedElements = Elements.newBuilder();
+    for (Map.Entry<String, Receiver<?>> entry : outputDataReceivers.entrySet()) {
+      ByteString bytes = entry.getValue().getOutput().toByteString();
+      if (bytes.isEmpty()) {
+        continue;
+      }
```

Comment:
Save on the object instantiation, ditto for timers.
```suggestion
      if (entry.getValue().getOutput().size() == 0) {
        continue;
      }
      ByteString bytes = entry.getValue().getOutput().toByteString();
```